sklearn/metrics/pairwise.py

```diff
@@ -552,15 +571,20 @@ def _euclidean_distances_upcast(X, XX=None, Y=None, YY=None, batch_size=None):
     X and Y are upcast to float64 by chunks, which size is chosen to limit
     memory increase by approximately 10% (at least 10MiB).
     """
+    xp, _, device_ = get_namespace_and_device(X, Y)
     n_samples_X = X.shape[0]
     n_samples_Y = Y.shape[0]
     n_features = X.shape[1]
 
-    distances = np.empty((n_samples_X, n_samples_Y), dtype=np.float32)
+    distances = xp.empty((n_samples_X, n_samples_Y), dtype=xp.float32, device=device_)
 
     if batch_size is None:
-        x_density = X.nnz / np.prod(X.shape) if issparse(X) else 1
-        y_density = Y.nnz / np.prod(Y.shape) if issparse(Y) else 1
+        x_density = (
+            X.nnz / np.prod(X.shape) if issparse(X) else xp.asarray(1, device=device_)
+        )
+        y_density = (
+            Y.nnz / np.prod(Y.shape) if issparse(Y) else xp.asarray(1, device=device_)
+        )
```

Comment:
why isn't this: 

```suggestion
            X.nnz / xp.prod(X.shape) if issparse(X) else xp.asarray(1, device=device_)
        )
        y_density = (
            Y.nnz / xp.prod(Y.shape) if issparse(Y) else xp.asarray(1, device=device_)
        )
```

Reply:
I think we can use xp as well.

Reply:
If `if issparse(Y)` then it's perfectly fine to use numpy because scipy sparse does not pretend to follow the array API (for the time being).

It's a matter of taste. I am fine with either option.